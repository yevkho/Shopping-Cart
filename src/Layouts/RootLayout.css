/* NAV BAR */
nav {
  padding: 10px 20px;

  display: flex;
  flex-direction: column;
  gap: 20px;
}

a {
  text-decoration: none;
  padding: 6px;
  border-radius: 4px;
  color: black;
}

a.active {
  background: var(--grey-color);
}

nav button {
  align-self: baseline;
  border: 1px solid transparent;
  padding: 1rem;
  cursor: pointer;
  border-radius: 50%;
}
nav button:hover {
  border-color: #646cff;
}

.cartSidebar {
  position: fixed;
  right: 0;
  top: 0;
  height: 100vh;
  width: 300px;
  background-color: white;
  box-shadow: -2px 0 5px rgba(0, 0, 0, 0.5);
  padding: 20px;
  /* z-index: 1000; */
  transform: translateX(100%); /* Initially hidden */
  transition: transform 0.3s ease-in-out;
}

.cartSidebar.open {
  transform: translateX(0); /* Slide in */
}

/* MAIN */
main {
  padding: 20px 30px;
}

.home,
.shop {
  display: flex;
  flex-direction: column;
  gap: 20px;
}

.ItemsContainer {
  padding: 10px 30px;

  display: grid;
  gap: 15px;
  grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
  grid-auto-rows: 1fr;
}

.ItemsContainer a {
  height: 24rem;

  padding: 10px;
  border: solid 1px black;
  border-radius: 5px;

  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: space-between;

  background-color: var(--grey-color);
  box-shadow: var(--boxshadow);

  transition: transform 0.2s;
}

.ItemsContainer a:hover {
  cursor: pointer;
  transform: scale(1.02);
}

.ItemsContainer img {
  height: 66%;
  object-fit: contain;
  max-width: 100%;
}
