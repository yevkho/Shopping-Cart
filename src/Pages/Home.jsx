export default function Home() {
  return (
    <>
      <div className="home">
        <h2>Home</h2>
        <p>Welcome to Home page!</p>
      </div>
    </>
  );
}
