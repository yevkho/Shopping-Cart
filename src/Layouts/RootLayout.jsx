import { useState } from "react";
import { NavLink, Outlet } from "react-router-dom";
import "../Layouts/RootLayout.css";

export default function RootLayout() {
  const [isCartOpen, setIsCartOpen] = useState(false);
  const [cartItems, setCartItems] = useState(0);

  function toggleCart() {
    setIsCartOpen(isCartOpen === true ? false : true);
  }

  return (
    <>
      <nav>
        <h3>Dashboard</h3>
        <NavLink to="/">Home</NavLink>
        <NavLink to="shop">Shop</NavLink>
        <button className="cartIcon" onClick={toggleCart}>
          Cart
        </button>
        <div>total items: {cartItems}</div>
      </nav>
      <main>
        <Outlet context={[setCartItems]} />
      </main>

      {isCartOpen && (
        <div
          className={`cartSidebar ${isCartOpen ? "open" : ""}`}
          data-testid="sidebar"
        >
          <button onClick={toggleCart}>Close</button>
          <h2>Your Cart</h2>
          <p>Items in cart: {cartItems}</p>
          <button onClick={() => alert("you are checked out!")}>
            Checkout
          </button>
        </div>
      )}
    </>
  );
}
