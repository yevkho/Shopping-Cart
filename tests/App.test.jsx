import { describe, it, expect } from "vitest";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MemoryRouter } from "react-router";

// import pages
import Home from "../src/Pages/Home";
// import Shop from "../src/Pages/Shop";
// import ShopItem from "../src/Pages/ShopItem";
// import ErrorPage from "../src/Pages/ErrorPage";
import RootLayout from "../src/Layouts/RootLayout";

// DASHBOARD
describe("Dashboard", () => {
  it("render 'a' Nav links correctly", () => {
    render(
      <MemoryRouter>
        {" "}
        <RootLayout />
      </MemoryRouter>
    );
    screen.debug();

    expect(screen.getByRole("link", { name: /home/i })).toBeInTheDocument;
    expect(screen.getByRole("link", { name: /shop/i })).toBeInTheDocument;
    expect(screen.getByRole("button", { name: "Cart" })).toBeInTheDocument;
  });

  it("Cart button click revealing side panel", async () => {
    const user = userEvent.setup();
    render(
      <MemoryRouter>
        {" "}
        <RootLayout />
      </MemoryRouter>
    );
    screen.debug();

    const button = screen.getByRole("button", { name: "Cart" });

    await user.click(button);

    expect(screen.getByTestId("sidebar")).toBeInTheDocument;
  });
});

//HOME
describe("Home", () => {
  it("renders headline 'Home'", () => {
    render(<Home />);
    screen.debug();

    // expect(screen.getByRole("heading")).toBeInTheDocument;
    expect(screen.getByRole("heading").textContent).toMatch("Home");
  });

  it("basic snapshot", () => {
    const { container } = render(<Home />);
    expect(container).toMatchSnapshot();
  });
});

//SHOP
// there is potential to mock test global.fetch(...) function
// with artificial resolution (e.g., 2-3) objects
// and see of they properly render on screen (e.g., https://github.com/NontasBak/shopping-cart/blob/main/src/components/Shop/Shop.test.jsx)
