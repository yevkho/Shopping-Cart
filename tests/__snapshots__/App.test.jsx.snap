// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App shallow testing > basic snapshot 1`] = `
<div>
  <h2>
    Shopping Cart
  </h2>
</div>
`;

exports[`Home > basic snapshot 1`] = `
<div>
  <div
    class="home"
  >
    <h2>
      Home
    </h2>
    <p>
      Welcome to Home page!
    </p>
  </div>
</div>
`;
